### DISCUSSION TITLE
Scheduling of upcoming PRs

### DISCUSSION THREAD

**Discussion by philpax** [rustformers/llama-rs#121] [General]
I've been merging PRs fairly haphazardly, but they're all starting to conflict with each other. I propose the following merge order and then *we ship 0.1.0*:

- [x] https://github.com/rustformers/llama-rs/pull/119 - most pressing, latest version of llama/ggml logic
- [x] https://github.com/rustformers/llama-rs/pull/122 - fixes byte token behaviour
- [x] https://github.com/rustformers/llama-rs/pull/116 - assuming that this is the correct fix
- [x] https://github.com/rustformers/llama-rs/pull/125 - blocked on getting it working, but this lays the foundation for loading models of other formats and should let us support all three GGML formats
- [x] https://github.com/rustformers/llama-rs/pull/84 - blocked on fixing the logic, but means users won't need upstream `llama.cpp` at all
- [ ] https://github.com/rustformers/llama-rs/pull/85 - finally, implements our first non-LLaMA model (with hopefully more to come), and will likely depend on loading logic touched by #114 and #84, and needs to be updated to match #119 and #116
- [x] and then cleanup work to integrate all of these together (esp with #84, which introduces loading helper functions that could be used by #114 and #85)

No other PRs that significantly touch logic will be merged outside of this order unless they're critical to getting things working.

Please confirm you're OK with this, or suggest an alternate plan: @setzer22 @KerfuffleV2 @iacore @FloppyDisck @hhamud

**Comment by KerfuffleV2** [comment]
It looks reasonable to me, but that's not worth much. I have no qualifications for juggling pull requests.

Is putting the GGML Rust bindings in a separate repo/project something that's on the roadmap? (I've been looking at trying to use it with my RWKV stuff for example, but that's going to require adding some ops and such.)

**Reply by philpax** [reply]
We might, yeah - separating `ggml-rs` out is a step towards that. For now, you should be able to use it as a Git dependency, but the additional ops would be a little annoying to manage. Let's see how we feel after the release.

**Reply by iacore** [reply]
`ggml` is not published to crates.io

you can use a specific git commit of this repo for now

**Comment by hhamud** [comment]
I'm fine with this

**Comment by KerfuffleV2** [comment]
By the way, #114 already works with a 1 line code change (and the nicer way to do it is obvious). Unless iacore doesn't like the idea of the non-mmap approach that one really just needs code cleanups to be suitable for merging.

**Comment by iacore** [comment]
Which PRs are considered "done"? Those should be merged first. I don't care about the order.

I want to fix the converting back and forth of `f16_` (should be named `tensor_element_type`) after this is all sorted out

**Comment by FloppyDisck** [comment]
It sounds good to me

**Comment by philpax** [comment]
Quick update on this: #84 is dependent on #125 in order to support all three model types (it currently only supports GGMJ), so we can't merge that yet. #85 is now *very* out of sync (sorry, @hhamud!) but hopefully the changes that have been introduced since will make it easier to reconstruct and to share as much code as possible.

**Reply by danforbes** [reply]
Given that some of the recent commits on this repository have been somewhat haphazard as well as the fact that there are many good/important ideas conflicting with each other, I wonder if now is a good time to think about using a separate `llm` repository as a place to take a more deliberate approach to the development of what could become an extremely important crate. I think it could be good to take a trait-first approach to the design of the `llm` crate, and follow some of the guidance provided by @setzer22 in [this comment](https://github.com/rustformers/llama-rs/pull/85#issuecomment-1508307588). It would also be nice to see a greater emphasis on testing, which seems like it would be very difficult to incorporate into this repository at this point. I would be more than happy to host a virtual design discussion if people are interested.

**Reply by philpax** [reply]
Our thinking here is that we'll get two models in (at this stage, LLaMA and BLOOM) and design a trait-based API for them, but the reason that hasn't happened yet is because many of the foundational elements are/were still in churn. Things are beginning to stabilise now, so we're in a better position to scope this out properly. That's also why we haven't released on `crates.io` yet - we're taking the time to make sure the first release can be used with all compatible models, and that its API is usable for a wide range of applications.

> It would also be nice to see a greater emphasis on testing, which seems like it would be very difficult to incorporate into this repository at this point.

That's something we're aware of, but there's no easy answers here - there's no way to test it without weights, and there are no weights for LLaMA that are small or legally unburdened enough to use in CI. The individual components can't really be tested in isolation without some form of weights, either.

Once/if there are LLaMA-arch compatible weights that can be used for deterministic testing, I imagine that this will improve significantly. Open to suggestions on what else can be done here.

**Reply by danforbes** [reply]
I took a stab at creating a `Model` trait and a `Llama` implementation...is this the kind of thing you had in mind? https://github.com/rustformers/llama-rs/compare/main...danforbes:llama-rs:dfo/refactor/traits

**Reply by hhamud** [reply]
> I took a stab at creating a `Model` trait and a `Llama` implementation...is this the kind of thing you had in mind? [main...danforbes:llama-rs:dfo/refactor/traits](https://github.com/rustformers/llama-rs/compare/main...danforbes:llama-rs:dfo/refactor/traits)

I had something quite similar in my bloom PR #85 where I attempted to implement a trait for both llama and the bloom models as they are quite similar, however may need something more flexible with incoming #75 model.

**Reply by danforbes** [reply]
Yes, I definitely referred to your changes! Unfortunately, they had fallen rather behind the tip of this repo, and I also tried to make fewer structural changes than you did.

**Reply by danforbes** [reply]
Here is a branch that incorporates the Bloom additions by @hhamud: https://github.com/rustformers/llama-rs/compare/main...danforbes:llama-rs:dfo/model/bloom
